File: photomap-core-api/api-gateway/src/main/java/ua/in/photomap/apigateway/filter/AuthenticationFilter.java
package ua.in.photomap.apigateway.filter;

import com.auth0.jwt.interfaces.DecodedJWT;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.cloud.context.config.annotation.RefreshScope;
import org.springframework.cloud.gateway.filter.GatewayFilter;
import org.springframework.cloud.gateway.filter.GatewayFilterChain;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;
import ua.in.photomap.apigateway.service.UserService;
import ua.in.photomap.apigateway.util.AuthUtils;
import ua.in.photomap.apigateway.util.CookieUtils;

import java.util.*;
import java.util.function.Supplier;

@RefreshScope
@Component
@RequiredArgsConstructor
@Slf4j
public class AuthenticationFilter implements GatewayFilter {
    private static final List<String> IGNORED_PATHS = Arrays.asList(
            "/api/auth/google",
            "/api/auth/login"
    );

    private final UserService userService;
    private final AuthUtils authUtils;
    private final CacheManager cacheManager;

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        ServerHttpRequest request = exchange.getRequest();
        log.debug("Auth filter start for path: {}", request.getPath());

        if (isIgnoredPath(request)) {
            return chain.filter(exchange);
        }

        Optional<String> authToken = authUtils.getAuthToken(request);
        if (authToken.isEmpty()) {
            return chain.filter(exchange);
        }

        return validateToken(exchange, chain, authToken.get());
    }

    private boolean isIgnoredPath(ServerHttpRequest request) {
        String path = request.getURI().getPath();
        return IGNORED_PATHS.stream()
                .anyMatch(path::contains);
    }

    private Mono<Void> validateToken(ServerWebExchange exchange, GatewayFilterChain chain, String token) {
        try {
            DecodedJWT jwt = authUtils.getDecodedJWT(token);
            Long userId = jwt.getClaim("user_id").asLong();
            Set<String> tokenPrivileges = new HashSet<>(jwt.getClaim("privileges").asList(String.class));

            log.debug("Validating auth for user: {}", userId);

            if (!validateUser(userId, tokenPrivileges)) {
                return onError(exchange, HttpStatus.FORBIDDEN);
            }

            return chain.filter(exchange);
        } catch (Exception e) {
            log.warn("Token validation failed: {}", e.getMessage());
            return onError(exchange, HttpStatus.FORBIDDEN);
        }
    }

    private boolean validateUser(Long userId, Set<String> tokenPrivileges) {
        if (!getIsEnabled(userId)) {
            log.debug("User {} is not enabled", userId);
            return false;
        }

        if (!getIsNonLockedUser(userId)) {
            log.debug("User {} is locked", userId);
            return false;
        }

        Set<String> userPrivileges = getUserPrivileges(userId);
        boolean privilegesMatch = Objects.equals(tokenPrivileges, userPrivileges);
        log.debug("User {} privileges match: {}", userId, privilegesMatch);

        return privilegesMatch;
    }

    private Set<String> getUserPrivileges(Long userId) {
        return getCachedValue("getUserPrivileges", userId, () -> {
            Set<String> privileges = userService.getUserPrivileges(userId);
            return !privileges.isEmpty() ? privileges : null;
        });
    }

    private boolean getIsNonLockedUser(Long userId) {
        Boolean result = getCachedValue("isNonLockedUser", userId, () -> {
            boolean nonLocked = userService.isNonLockedUser(userId);
            return nonLocked ? true : null;
        });
        return result != null && result;
    }

    private boolean getIsEnabled(Long userId) {
        Boolean result = getCachedValue("isEnabledUser", userId, () -> {
            boolean enabled = userService.isEnabledUser(userId);
            return enabled ? true : null;
        });
        return result != null && result;
    }

    private <T> T getCachedValue(String cacheName, Long userId, Supplier<T> valueSupplier) {
        Cache cache = cacheManager.getCache(cacheName);
        if (cache == null) {
            log.warn("Cache {} not found", cacheName);
            return valueSupplier.get();
        }

        T cachedResult = cache.get(userId, (Class<T>) Object.class);
        if (cachedResult != null) {
            log.debug("Cache hit for {}: {}", cacheName, cachedResult);
            return cachedResult;
        }

        T result = valueSupplier.get();
        if (result != null) {
            cache.put(userId, result);
            log.debug("{} value cached for user {}", cacheName, userId);
        }

        return result;
    }

    private Mono<Void> onError(ServerWebExchange exchange, HttpStatus status) {
        ServerHttpResponse response = exchange.getResponse();
        response.setStatusCode(status);
        CookieUtils.clearCookie(response, "accessToken");
        return response.setComplete();
    }
}

File: photomap-core-api/api-gateway-kt/src/main/kotlin/ua/in/photomap/apigateway/filter/AuthenticationFilter.kt
package ua.`in`.photomap.apigateway.filter

import com.auth0.jwt.interfaces.DecodedJWT
import org.slf4j.LoggerFactory
import org.springframework.cache.CacheManager
import org.springframework.cloud.gateway.filter.GatewayFilter
import org.springframework.cloud.gateway.filter.GatewayFilterChain
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import ua.`in`.photomap.apigateway.service.UserService
import ua.`in`.photomap.apigateway.util.AuthUtils
import ua.`in`.photomap.apigateway.util.CookieUtils

@Component
class AuthenticationFilter(
    private val userService: UserService,
    private val authUtils: AuthUtils,
    private val cacheManager: CacheManager
) : GatewayFilter {

    companion object {
        private val log = LoggerFactory.getLogger(AuthenticationFilter::class.java)
        private val IGNORED_PATHS = listOf(
            "/api/auth/google",
            "/api/auth/login"
        )
    }

    override fun filter(exchange: ServerWebExchange, chain: GatewayFilterChain): Mono<Void> {
        log.debug("Auth filter start.")
        val request = exchange.request

        if (!requiresAuthentication(request.uri.path)) {
            return chain.filter(exchange)
        }

        return try {
            authUtils.getAuthToken(request)?.let { token ->
                val jwt = authUtils.getDecodedJWT(token)
                validateToken(exchange, chain, jwt)
            } ?: onError(exchange, HttpStatus.FORBIDDEN)
        } catch (e: Exception) {
            log.warn("Auth filter caught exception: {}", e.message)
            onError(exchange, HttpStatus.FORBIDDEN)
        }
    }

    private fun validateToken(
        exchange: ServerWebExchange,
        chain: GatewayFilterChain,
        jwt: DecodedJWT
    ): Mono<Void> {
        val userId = jwt.getClaim("user_id").asLong()
        val tokenPrivileges = jwt.getClaim("privileges").asList(String::class.java).toSet()

        log.debug("Start auth check. User id: {}", userId)

        return when {
            !getIsEnabled(userId) -> {
                log.debug("User {} is not enabled", userId)
                onError(exchange, HttpStatus.I_AM_A_TEAPOT)
            }
            !validateUserAccess(userId, tokenPrivileges) -> {
                log.debug("User {} access validation failed", userId)
                onError(exchange, HttpStatus.FORBIDDEN)
            }
            else -> chain.filter(exchange)
        }
    }

    private fun validateUserAccess(userId: Long, tokenPrivileges: Set<String>): Boolean {
        val isNonLocked = getIsNonLockedUser(userId)
        log.debug("User {} is non locked: {}", userId, isNonLocked)
        
        if (!isNonLocked) return false
        
        val userPrivileges = getUserPrivileges(userId)
        log.debug("User {} privileges match: {}", userId, tokenPrivileges == userPrivileges)
        
        return tokenPrivileges == userPrivileges
    }

    private fun requiresAuthentication(path: String): Boolean =
        IGNORED_PATHS.none { path.contains(it) }

    private fun getIsEnabled(userId: Long): Boolean =
        getCachedValue("isEnabledUser", userId) {
            userService.isEnabledUser(userId)
        } ?: false

    private fun getIsNonLockedUser(userId: Long): Boolean =
        getCachedValue("isNonLockedUser", userId) {
            userService.isNonLockedUser(userId)
        } ?: false

    private fun getUserPrivileges(userId: Long): Set<String> =
        getCachedValue("getUserPrivileges", userId) {
            userService.getUserPrivileges(userId)
        } ?: emptySet()

    private fun <T> getCachedValue(cacheName: String, key: Any, supplier: () -> T): T? {
        val cache = cacheManager.getCache(cacheName) ?: return supplier()
        
        return cache.get(key)?.get() as? T ?: supplier().also { result ->
            if (result != null) {
                cache.put(key, result)
                log.debug("{} value cached for key {}", cacheName, key)
            }
        }
    }

    private fun onError(exchange: ServerWebExchange, status: HttpStatus): Mono<Void> {
        val response = exchange.response
        response.statusCode = status
        CookieUtils.clearCookie(response, "accessToken")
        return response.setComplete()
    }
} 